src/gpio.rs

```diff
@@ -236,6 +236,29 @@ where
     pub fn inner(&self) -> &P {
         &self.inner
     }
+
+    /// Indicates wether the voltage at the pin is currently HIGH
+    ///
+    /// This can be used when the pin is in any direction:
+    ///
+    /// If it is currently an Output pin, it indicates whether the pin output is set to HIGH
+    /// If it is currently an Input pin, it indicates wether the pin input is HIGH
+    ///
+    /// This method is only available, if two conditions are met:
+    /// - The pin is in the GPIO state.
+    /// - The pin direction is set to input.
+    ///
+    /// See [`Pin::into_input_pin`] and [`into_input`]. Unless both of these
+    /// conditions are met, code trying to call this method will not compile.
+    ///
+    /// [`Pin::into_input_pin`]: ../pins/struct.Pin.html#method.into_input_pin
+    /// [`into_input`]: #method.into_input
+    pub fn is_high(&self) -> bool {
+        let gpio = unsafe { &*pac::GPIO::ptr() };
+        let registers = Registers::new(gpio);
+
+        is_high(&registers, self.inner())
+    }
```

Comment:
This method seems have wandered into the unconditional `impl` block, meaning it is now available regardless of pin direction. That doesn't seem right. Shouldn't it just stay where it is?

Reply:
we can't keep it where it is, because we need to be able to access it from `Level:: from_pin` (or we'd need 3 implementations, for `D: Direction::Input`, `D: Direction::Output` and `D: Direction::Dynamic`.
*However*, I'd argue that this is is not a bug, but actually a feature– maybe you'd want to read the level of an Output pin, to assert that it behaves as you're expecting it to. That'd also enable us to just get rid of the awkwardly-named `get_set_level` … I don't know why I didn't already do that

Reply:
Ah, okay, I understand. The method is available regardless of state, after all, as there's no way to have an uninitialized `GpioPin`.

So I guess we have to address the following questions/issues:
- Do we want to make the formerly input-only methods available unconditionally? That seems attractive to me, due to the simplicity and because it enables `Level::from_pin`.
- If yes, we should be consistent about it and do this for all input methods (`is_high`, `is_low`, `get_level`).
- If yes, we need to update the documentation, as that's currently wrong.
- If yes, do we want to keep `is_set_high`/`is_set_low`? They would be redundant, but a case could be made that they should be kept, to follow the style of the embedded-hal traits.

@david-sawatzke, what do you think? I'm currently tending towards making `is_high`/`is_low`/`get_level` unconditional and removing `is_set_high`/`is_set_low`, but I'm somewhat undecided.

Reply:
(This is more of a drive-by comment, I haven't thought deeply about this)

I'd tend to keep them both separate, for a couple of reasons:
1. Consistency with the HAL traits
2. If (or when) we add an OD function, they need to be separate
3. Reduce confusion, imagine e.g. having a high load on an output pin and hoping `is_high` could be used to detect it.

I'd also not expose them for output-only pins, so it becomes harder to miss accidentally using a pin in the wrong direction.

Maybe as an alternative then add a `pub(crate)` `fn` unconditionally and implement `Level::form_pin` with that for both input & output (not sure if that's more easily said than done)?

Reply:
Hmm, yeah, makes sense.

> Maybe as an alternative then add a `pub(crate)` `fn` unconditionally and implement `Level::form_pin` with that for both input & output (not sure if that's more easily said than done)?

Seems like a workable solution. What do you think, @Lotterleben?

Reply:
that sounds really good! will look into it tomorrow.